src/reducers/viewconfig.ts

```diff
@@ -0,0 +1,302 @@
+                                        reference: 'clusius_Persons'
+                                    }, {
+                                        value: 'tim_gender',
+                                    }]
+                                }
+                            }
+                        ]
+                    }
+                ]
+            }
+        ]
+    },
+    {
+        type: COMPONENTS.keyValue,
+        key: {
+            field: 'to'
+        },
+        values: [
+            {
+                type: COMPONENTS.value,
+                value: {
+                    fields: [{
+                        value: 'tim_hasResident',
+                        reference: 'clusius_Persons'
+                    }, {
+                        value: 'tim_hasBirthPlace',
+                        reference: 'clusius_Places'
+                    }, {
+                        value: 'tim_country',
+                    }]
+                }
+            }
+        ]
+    }
+];
+
+const initialState: ViewConfigReducer = normalizeTree(exampleData);
+
+// action definitions
+
+type AddViewConfigNodeAction = {
+    type: 'ADD_VIEW_CONFIG_NODE',
+    payload: {
+        component: Component
+    }
+};
+
+type ModifyViewConfigNodeAction = {
+    type: 'MODIFY_VIEW_CONFIG_NODE',
+    payload: {
+        nodeId: number,
+        component: Component
+    }
+};
+
+type ViewConfigChildAction = {
+    type: 'ADD_VIEW_CONFIG_CHILD' | 'DELETE_VIEW_CONFIG_CHILD',
+    payload: {
+        nodeId: number,
+        childId: number
+    }
+};
+
+type SortViewConfigChildAction = {
+    type: 'SORT_VIEW_CONFIG_CHILD',
+    payload: {
+        nodeId: number,
+        oldIndex: number,
+        newIndex: number
+    }
+};
+
+type DeleteViewConfigItemAction = {
+    type: 'DELETE_VIEW_CONFIG_NODE',
+    payload: {
+        nodeId: number
+    }
+};
+
+type Action =
+    AddViewConfigNodeAction
+    | DeleteViewConfigItemAction
+    | ViewConfigChildAction
+    | ModifyViewConfigNodeAction
+    | SortViewConfigChildAction;
+
+// selectors
+
+export const getNodeById = (id: number, state: ViewConfigReducer): NormalizedComponent | undefined =>
+    state.find(item => item.id === id);
+
+export const lastId = (state: ViewConfigReducer): number => state
+    .map(item => item.id)
+    .reduce(((previousValue: number, currentValue: number) => Math.max(previousValue, currentValue)), -Infinity);
+
+const getAllDescendantIds = (state: ViewConfigReducer, nodeId: number) => (
+    getNodeById(nodeId, state)!.childIds.reduce((acc, childId) => [...acc, childId, ...getAllDescendantIds(state, childId)], [])
+);
+
+const deleteMany = (state: ViewConfigReducer, ids: number[]): ViewConfigReducer => {
+    state = [...state];
+    for (const id of ids) {
+        const index = state.findIndex(item => item.id === id);
+        state.splice(index, 1);
+    }
+    return state;
+};
+
+// reducers
+
+const childIds = (state, action) => {
```

Comment:
Can you add typings, so it is easier to understand what is expected to actually be the state?